xwiki-platform-core/xwiki-platform-webjars/xwiki-platform-webjars-api/src/main/java/org/xwiki/webjars/internal/WebJarsResourceReferenceHandler.java

```diff
@@ -120,4 +115,22 @@ protected ClassLoader getClassLoader(String namespace)
     {
         return this.classLoaderManager.getURLClassLoader(namespace, true);
     }
+
+    @Override
+    protected String getContentType(InputStream resourceStream,
+        WebJarsResourceReference resourceReference) throws IOException
+    {
+        String mimeType;
+        // When the resource is not cacheable and the request resource is a less file, we compile it to css.
+        // In this case, the content type must be explicitly set to text/css instead of the default text/x-less
+        // that would otherwise be computed from the resource filename.
```

Comment:
I guess you mean by Tika because `getContentType()` should guess the right value since it looks at both the name and the content. Or maybe rewrite it the comment as:

> In this case, we set the content type text/css explicitly in order to be safe since otherwise it's possible that the default implementation of getContentType() will infer the type only by looking at the resource file name extension and not check its content (or fail to discover the right content).

?

Reply:
BTW this makes it even more important that the javadoc of getContentType() mentions that it should be overridden when the type is known (see https://github.com/xwiki/xwiki-platform/pull/1586/files#r607793345)?

Reply:
> getContentType() should guess the right value since it looks at both the name and the content.

Not in this case. Tika guess the content only for types which use a magic marker at the beginning, I don't think it ever tries to guess what's in a text file.

Reply:
@vmassol I don't see any problem with @manuelleduc 's comment. The resource is **transformed** (LESS -> CSS) and the default implementation of ``getContentType()`` returns the content type of the **original** resource (LESS), before the transformation. We need the content type to match the result of the transformation (CSS).

Reply:
> important that the javadoc of getContentType() mentions that it should be overridden when the type is known

No, you need to overwrite ``getContentType()`` only if you transform the resource, i.e. you change it's content type. Which is what happens here. We change the content type form LESS to CSS. The resource file in the WebJar has content type LESS while we produce CSS because we transform the resource.

Reply:
@mflorea The javadoc for `getContentType()` says:

> * Computes the content type of the resource. By default the content type is inferred by {@link TikaUtils#detect(InputStream, String)} based on the resource content and name.

So it's supposed to infer the type not just from the name but also from the resource content according to the javadoc.

Reply:
> No, you need to overwrite getContentType() only if you transform the resource, 

I disagree with this. It's always better to set it if you know it, you'll avoid the cost of Tika trying to guess it.

Reply:
> Not in this case. Tika guess the content only for types which use a magic marker at the beginning, I don't think it ever tries to guess what's in a text file.

@tmortagne ok but still, I think my comment is better and takes into account different default implementations (you cannot guarantee that Tika in the future will not check the content for ex. or that we'll stop using Tika and guess it ourselves).

Reply:
> So it's supposed to infer the type not just from the name but also from the resource content according to the javadoc.

And? The content is LESS but we **transform the content**. So the content type determined based on the content (which is LESS) can't be right in this case (we output CSS), which is why we need to overwrite ``getContentType()``.

Reply:
> And? The content is LESS but we transform the content. So the content type determined based on the content (which is LESS) can't be right in this case (we output CSS), which is why we need to overwrite getContentType().

ok, indeed, that's a good reason :) Now we still have this small perf disagreement but it's orthogonal.

Reply:
> It's always better to set it if you know it

You "know it" only if you need to filter / transform the resource (i.e. to handle it in a special way). There's no need to overwrite ``getContentType`` otherwise (we would duplicate Tika's code).